### DISCUSSION TITLE
Remote database

### DISCUSSION THREAD

**Discussion by mdre** [ArcadeData/arcadedb#115] [General]
With RemoteDatabase you connect to a db:

RemoteDatabase arcadeDB = new RemoteDatabase("localhost",2480,"ogm-test","root","rootroot");

but, is the database open? how should i start a transaction?

The local Database object has open() and transaction() method but i don't see it in RemoteDatabase.

Thank's

**Comment by mdre** [comment]
Sorry.  Forget this! 

db.begin()
..
db.commit() 

works.

**Comment by lvca** [comment]
@mdre with the latest release we removed begin() and commit() from the remote database class because transactions on the client side are not managed correctly. The server is stateless, so we'd need to bind the HTTP session to a server-side session.

For now, each operation is atomic. If you need to execute multiple commands in a transaction, the only way is to execute a command as a script with `begin;<your commands separated by ;>commit;`

**Reply by mdre** [reply]
Wow... it would be quite a challenge to rethink a system to operate that way.
I'm going to try but it doesn't seem like it's that simple to replace Orient with Arcade.

**Reply by tetious** [reply]
I've been working on a simple C# wrapper for the HTTP API, and figuring out transactions is an interesting problem. 

If you think of transactions as batches, and use the [sql batch](https://docs.arcadedb.com/#SQL-Batch) stuff cleverly, I think it will fit in many cases.

I'm just not sure yet how to make actually using it ergonomic and familiar to users used to stateful transactions and multiple result sets within a given transaction. Maybe it is swimming upstream to use a stateless protocol like http for something like this. Using something like gRPC might be interesting.

Of course, the other driver options may not have this problem. I haven't yet experimented with those yet.

**Reply by lvca** [reply]
I think the easiest thing to do is to bind a client session to a server-side session. On the server side we could create a session id as a UUID and send it to the client at authentication as a token. If the client passes this token in the HTTP request header, then the server looks for the session id. In this way, we could save a transaction object and expose the begin + commit commands in the remote database API.
